pkg/port/port.go

```diff
@@ -236,6 +240,86 @@ func GetConnections(execClient exec.Client, podName string, containerName string
 	return connections, nil
 }
 
+// CheckAppPortsListening checks whether all the specified ports are really opened and in LISTEN mode in each corresponding container
+// of the pod specified.
+// It does so by periodically looking inside the container for listening connections until it finds each of the specified ports,
+// or until the specified timeout has elapsed.
+func CheckAppPortsListening(
+	ctx context.Context,
+	execClient exec.Client,
+	podName string,
+	containerPortMapping map[string][]int,
+	timeout time.Duration,
+) error {
+	if len(containerPortMapping) == 0 {
+		return nil
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	hasPortFn := func(connections []Connection, p int) bool {
+		for _, c := range connections {
+			if p == c.LocalPort {
+				return true
+			}
+		}
+		return false
+	}
+
+	g := new(errgroup.Group)
+	for container, ports := range containerPortMapping {
+		container := container
+		ports := ports
+
+		g.Go(func() error {
+			b := backo.NewBacko(1*time.Second, 2, 0, 10*time.Second)
+			ticker := b.NewTicker()
+			portsNotListening := make(map[int]struct{})
+
+			for {
+				select {
+				case <-ctxWithTimeout.Done():
+					var unopenedPorts []string
+					for p := range portsNotListening {
+						unopenedPorts = append(unopenedPorts, strconv.Itoa(p))
+					}
+					msg := fmt.Sprintf("timeout while checking for ports in container %q", container)
```

Comment:
In case of multiple containers, this will only return the current container with unreachable port. Perhaps we should return an error message container all the containers and their unreachable port?

In the output below, it only prints about runtime:3030, when it could also print info about tools:5000.
```sh
 ✗  Waiting for the application to be ready [1m]
 ⚠  Port Forwarding might not work correctly: timeout while checking for ports in container "runtime"; ports not listening: 3030: context deadline exceeded
 -  Forwarding from 127.0.0.1:20001 -> 5000

 -  Forwarding from 127.0.0.1:20002 -> 3000

 -  Forwarding from 127.0.0.1:20003 -> 3030
```


```yaml
- container:
    endpoints:
    - name: http-node
      targetPort: 3000
    - exposure: none
      name: debug
      targetPort: 5858
    - name: http-3030
      targetPort: 3030
  name: runtime
- container:
    endpoints:
    - name: http-5000
      targetPort: 5000
    - name: debug-node
      targetPort: 8000
  name: tools
```

Reply:
Good catch. I initially wanted to return early as soon as the timeout is reached, but I agree it makes sense, as all the goroutines would also have finished anyways. So we could collect all the unreachable ports and containers. I'll add it.

Reply:
Done - see [`55b6fd7` (#6701)](https://github.com/redhat-developer/odo/pull/6701/commits/55b6fd7a5f7a2456321973db42c9a993737f7e58)